### ISSUE TITLE
Initial Release/Development and Integration Plans

### ISSUE THREAD

**Issue by Tokazama** [JuliaNeuroscience/NeuroCore.jl#4]
I've been doing a lot of work for this around the Julia community and wanted to make sure that if anyone was following this project they knew something was actually happening. Here's the plan for getting this package released (probably an alpha release where things could change drastically given feedback) and properly integrated. 

## Initial Release/Development

* [ ] [This branch](https://github.com/JuliaNeuroscience/NeuroCore.jl/tree/pre-release) is the current pre-release. And is the barebones implementation to getting up and running. For those curious this has been a result of:
  - Better property creation and documentation: I developed [FieldProperties.jl](https://github.com/Tokazama/FieldProperties.jl) so that we could have a standard set of syntax that has documentation and optionally enforces specific return types and potential defaults. I'll document the benefits of this more elsewhere, but the most immediately noticeable difference is how much it cuts down on lines of could.
  - Eliminate properties that were either excessively specific or their utility in this package was highly uncertain. I included a lot of them in the initial PR so they could at least be in the commit history if they need to be obtained again.
* [ ] Comprehensive testing: This is a core package that I anticipate very few new types being a part of. Therefore, I anticipated code coverage to be crucial and measured accurately by Codecov measures.
  - [ ] Type stability: Where appropriate we should ensure that users don't have to worry about NeuroCore.jl causing type stability issues
  - [ ] Sensible defaults: I don't anticipate many to be implemented. Right now the default is to error unless it's part of a dictionary look-up for the property.
  - [ ] User level named dimension manipulation: There is some important user level functionality dedicated to manipulating named dimensions that are important to neuroscience data.
* [ ] Documentation:
  - [ ] Thoroughly document all user facing methods
  - [ ] Formal documentation manual
  - [ ] Readme/introduction with intention/purposes/goals of NeuroCore.jl

## Other Ecosystem Integration

Other JuliaNeuroscience packages are being addressed here #3. As for working with other ecosystems:

* [ ] I would really like other I/O packages to be able to use this as a common method of accessing they're metadata so users don't have to worry about learning a new set of methods for every file type. Ideally the type structure won't have to change too much, but the method of accessing data inside it should change to be consistent. I'll start by finishing up a revamp of NIfTI.jl and then reach out to other package maintainers. Of course, this will all depend on how well these suggestions are received 🤞.
* [ ] Move some functions to other packages and take on as dependencies: Some methods are specific to a certain type of analysis (e.g., time series analysis). Where it is appropriate I'd like these methods to move to their proper home at some point. Unfortunately, that doesn't exist yet.
* [ ] Support ecosystems whose work is important to the success of neuroscience research in Julia. Here's a tentative list:
  - JuliaImages
  - JuliaPlots
  - JuliaStats
  - JuliaArrays
  - JuliaDSP

**Comment by timholy** [comment]
WRT an AxisArrays replacement, should we evaluate [DimensionalData.jl](https://github.com/rafaqz/DimensionalData.jl)? Both DimensionalData and NamedDims.jl happen to be standing at 227 commits, but the current effort being spent in DD is quite high.

**Comment by Tokazama** [comment]
> should we evaluate DimensionalData.jl?

tl;dr I'm open to whatever solution the wider community ends up adopting (particularly JuliaImages) but I currently think that NamedDims has some big advantages

Although DimensionalData.jl has good performance I think NamedDims.jl has had thought put into every single method's performance. For example you'll find this comment peppered throughout the code `0-Allocations see: `@btime  (()->dim((:a, :b), :b))()`. Part of this may just be that more people seem to be contributing to NameDims.jl in discussion and PRs.

I have a lot of nit picky things about its design that I don't like. Most of my issue with it stem from it requiring users to specify dedicated types for each dimension. I think this results in a lot of unfriendly syntax and I suspect that it could lead to more burdensome maintenance in the future if it's widely adopted (if two different packages define the `Time` dimension type then everything breaks when those two packages are loaded).

I admit that a lot of these reasons aren't completely concrete (ugly syntax to one is beautiful Python syntax to another 😉). Therefore, if the consensus ends being that DimensionalData is the way to go then I will fully support that.

One last point. I think the problem with indexing could be solved relatively quickly once people agree and get behind a single approach. I really like what @mcabbott and I are circling around at the end of https://github.com/JuliaCollections/AxisArraysFuture/issues/1. I have most of the code implemented for what I've proposed and I'm starting to right some more examples [here](https://tokazama.github.io/StaticRanges.jl/dev/abstractaxis/).

**Comment by timholy** [comment]
Good to know. Until this morning I hadn't realized I wasn't watching NamedDims.jl, so I missed all those discussions. Maybe a good thing though, I've not been ready yet to tackle this issue with the seriousness it deserves, so best if others take the lead. But some of our other cleanups are getting sufficiently complete that this one is rising higher in the priority queue.

**Comment by Tokazama** [comment]
I just pushed some more examples and benchmarks to the last link I referenced. It's not pretty enough for a formal package, but I've tried to explore a lot of generic array interfaces as far as I can reasonably take it (e.g., how would it look to perform `cat` `append!`, `push!`, etc.). The trouble I've found with a lot of this is it's hard to explain why a certain approach might be better without getting a working demonstration, so hopefully this will prove helpful as people have time to look at it.

> I've not been ready yet to tackle this issue with the seriousness it deserves

I appreciate any contributions you're able to make. I understand you are very busy with other things.

As for the rest of the roadmap (or lack thereof) I will be finish the broad picture over the next couple days and create separate issues to explore the more granular details.

**Comment by Tokazama** [comment]
Registering a package soon that should help with (or even sole) the array indexing issue here: https://github.com/Tokazama/AxisIndices.jl

**Comment by behinger** [comment]
not sure where the best place is to write this, I wrote a very basic eeglab .set/.fdt importer:
https://github.com/unfoldtoolbox/unfold.jl/blob/master/test/debug_readEEGlab.jl

Needs to be adapted and extended to NeuroCore - I will do so happily once a bit more docu is there :)
